core/src/cluster_info.rs

```diff
@@ -1951,6 +1939,62 @@ impl ClusterInfo {
         Some(packets)
     }
 
+    fn handle_batch_pull_responses(
+        &self,
+        responses: Vec<(Pubkey, Vec<CrdsValue>)>,
+        thread_pool: &ThreadPool,
+        stakes: &HashMap<Pubkey, u64>,
+        epoch_time_ms: u64,
+    ) {
+        type HashMap = std::collections::HashMap<Pubkey, Vec<CrdsValue>>;
+        if responses.is_empty() {
+            return;
+        }
+        fn extend(hash_map: &mut HashMap, (from, mut data): (Pubkey, Vec<CrdsValue>)) {
+            match hash_map.entry(from) {
+                Entry::Occupied(mut entry) => {
+                    let value = entry.get_mut();
+                    if value.len() < data.len() {
+                        std::mem::swap(value, &mut data);
+                    }
+                    value.extend(data);
+                }
+                Entry::Vacant(entry) => {
+                    entry.insert(data);
+                }
+            }
+        }
+        fn merge(mut hash_map: HashMap, other: HashMap) -> HashMap {
+            if hash_map.len() < other.len() {
+                return merge(other, hash_map);
+            }
+            for kv in other {
+                extend(&mut hash_map, kv);
+            }
+            hash_map
+        }
+        let responses = thread_pool.install(|| {
+            responses
+                .into_par_iter()
+                .with_min_len(1024)
```

Comment:
This is an interesting way to join the responses together faster. Do we know how large the "responses" vector typically is? Wondering if it's enough to take advantage of these 1024-len chunks

Reply:
There are these two metrics:
https://github.com/solana-labs/solana/blob/e6b821c39/core/src/cluster_info.rs#L1997-L2000
in the following function, which may give an approximate _lower-bound_ for the length of responses here. Based on that approximate, on "average", this particular instance probably will not do much parallel processing today on mainnet.

But, nevertheless, I think it is good to do this using parallel iterators here, because it will help this scale should the network grow much bigger. It will also keep things more robust should some nodes fall behind and receive abnormal traffic. Even today on mainnet,  99% percentile is more than twice the average (and "max" is so far out).

![cluster_info_stats3 process_pull_response_len_tail](https://user-images.githubusercontent.com/1288998/96805054-38c62f80-1400-11eb-8b8c-79e61ecd1007.png)